File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list->
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        
        ListNode* prev_node = nullptr;
        ListNode* cur = head;
        auto index = 1;
        while (cur != nullptr){
            if (index == n+1){
                prev_node = head;
            }
            else if (index > n+1){
                prev_node = prev_node->next;
            }
            cur = cur->next;
            index += 1;
        }
        // if removed node is head
        if (prev_node == nullptr){
            auto new_head = head->next;
            delete head;
            return new_head;
        }

        //remove node if node is not head
        auto deleted_node = prev_node->next;
        prev_node->next = deleted_node->next;
        delete deleted_node;
        return head;

    }
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode()
        dummy.next = head

        # find n+1th node form the end of the list
        # use slow, fast pointer
        slow = dummy
        fast = dummy
        
        for _ in range(n + 1):
            fast = fast.next
        
        while fast:
            fast = fast.next
            slow = slow.next

        #remove nth node ( n+1th node next = n - 1 th node )
        slow.next = slow.next.next

        # return head
        return dummy.next